src/viam/sdk/components/camera.cpp

```diff
@@ -24,6 +25,104 @@ API API::traits<Camera>::api() {
+// Intended to be used in a sequential manner.
+template <typename T>
+T read_big_endian(const std::vector<unsigned char>& data, size_t* offset) {
+    if (data.size() < *offset + sizeof(T)) {
+        throw std::runtime_error("Attempted to read beyond data bounds.");
+    }
+
+    T value;
+    std::memcpy(&value, &data[*offset], sizeof(T));
+    value = boost::endian::big_to_native(value);
+    *offset += sizeof(T);
+
+    return value;
+}
+
+}  // namespace
+
+std::vector<unsigned char> Camera::encode_depth_map(const Camera::depth_map& m) {
+    const auto& shape = m.shape();
+    if (shape.size() != 2) {
+        throw Exception("Depth map is not a 2D array.");
+    }
+
+    const size_t height = shape[0];
+    const size_t width = shape[1];
+    const size_t total_byte_count = k_header_size + height * width * sizeof(uint16_t);
+    std::vector<unsigned char> data(total_byte_count);
+    size_t offset = 0;
+
+    // Network data is stored in big-endian, while most host systems are little endian.
+    append_big_endian(data, &offset, k_magic_number);
+    append_big_endian(data, &offset, width);
+    append_big_endian(data, &offset, height);
+
+    for (size_t i = 0; i < height; ++i) {
+        for (size_t j = 0; j < width; ++j) {
+            append_big_endian(data, &offset, m(i, j));
+        }
+    }
+
+    return data;
+}
+
+Camera::depth_map Camera::decode_depth_map(const std::vector<unsigned char>& data) {
+    if (data.size() < k_header_size) {
+        throw Exception("Data too short to contain valid depth information. Size: " +
+                        std::to_string(data.size()));
+    }
+
+    size_t offset = 0;
+    const uint64_t magic_number = read_big_endian<uint64_t>(data, &offset);
+    if (magic_number != k_magic_number) {
+        throw Exception(
+            "Invalid header for a vnd.viam.dep encoded depth image. The data may be corrupted, or "
+            "is not a Viam-encoded depth map.");
+    }
+
+    const uint64_t width = read_big_endian<uint64_t>(data, &offset);
+    const uint64_t height = read_big_endian<uint64_t>(data, &offset);
+
+    const auto expected_size = k_header_size + width * height * sizeof(uint16_t);
+    if (data.size() != expected_size) {
+        throw Exception("Data size does not match width, height, and depth values. Actual size: " +
+                        std::to_string(data.size()) +
+                        ". Expected size: " + std::to_string(expected_size) +
+                        ". Width: " + std::to_string(width) + " Height: " + std::to_string(height));
+    }
+
+    std::vector<uint16_t> depth_values;
+    depth_values.reserve(width * height);
+    for (size_t i = 0; i < width * height; ++i) {
+        depth_values.push_back(read_big_endian<uint16_t>(data, &offset));
+    }
+
+    xt::xarray<uint16_t> m = xt::xarray<uint16_t>::from_shape({height, width});
+    std::copy(depth_values.begin(), depth_values.end(), m.begin());
```

Comment:
I looked up a method to copy the memory in. Let me know if there's a better way to do it @acmorrow

Reply:
Hey, I was actually just looking at that. So, what you have here is fine, though it does cost you a copy. If you wanted to avoid the copy though, I think that's totally doable.

The trick would be to not create a `std::vector`, but just to directly create the `xt::xarray` you want to return, of the appropriate size and shape, before the loop, and then populate it in the loop.

```
xt::xarray<uint16_t> m = xt::xarray<uint16_t>::from_shape({height, width});
for (size_t i = 0; i < width * height; ++i) {
    m(<ii>, <jj>) = read_big_endian<uint16_t>(data, &offset);
```

Now, that'd require a little hassle to compute the correct `ii` and `jj` w.r.t. `height` and `width` and `i`, which is sort of a hassle. However, `xtensor` lets you reshape and view. So I think you could do it something like this:

```
xt::xarray<uint16_t> m = xt::xarray<uint16_t>::from_shape({height, width});
auto m_linear_view = xt::flatten(m);
for (size_t i = 0; i < width * height; ++i) {
    m_linear_view[i] = read_big_endian<uint16_t>(data, &offset));
return m;
```

It may not be exactly `flatten`, you might need `reshape_view`? I'd need to spend a little time with the `xtensor` docs to be sure. But something from https://xtensor.readthedocs.io/en/latest/view.html.

I'd say spend no more than 15 minutes on it. If you can make it work, great. If not, put in a TODO and leave it as a copy.